diff --git a/compiler/GHC/Driver/Main.hs b/compiler/GHC/Driver/Main.hs
index fd7b0c9..cb7d6b0 100644
--- a/compiler/GHC/Driver/Main.hs
+++ b/compiler/GHC/Driver/Main.hs
@@ -773,15 +773,19 @@ tcRnModule' sum save_rn_syntax mod = do
 -- | Convert a typechecked module to Core
 hscDesugar :: HscEnv -> ModSummary -> TcGblEnv -> IO ModGuts
 hscDesugar hsc_env mod_summary tc_result =
-    runHsc hsc_env $ hscDesugar' (ms_location mod_summary) tc_result
+    runHsc hsc_env $ hscDesugar' (Just mod_summary) (ms_location mod_summary) tc_result
 
-hscDesugar' :: ModLocation -> TcGblEnv -> Hsc ModGuts
-hscDesugar' mod_location tc_result = do
+hscDesugar' :: Maybe ModSummary -> ModLocation -> TcGblEnv -> Hsc ModGuts
+hscDesugar' mod_summary mod_location tc_result = do
     hsc_env <- getHscEnv
-    ioMsgMaybe $ hoistDsMessage $
+    r <- ioMsgMaybe $ hoistDsMessage $
       {-# SCC "deSugar" #-}
       deSugar hsc_env mod_location tc_result
 
+    -- Run plugins' desugar actions on the freshly desugared Core.
+    withPlugins (hsc_plugins hsc_env)
+      (\p opts mod_guts -> desugarResultAction p opts mod_summary tc_result mod_guts) r
+
 -- | Make a 'ModDetails' from the results of typechecking. Used when
 -- typechecking only, as opposed to full compilation.
 makeSimpleDetails :: Logger -> TcGblEnv -> IO ModDetails
@@ -1111,7 +1115,7 @@ hscDesugarAndSimplify summary (FrontendTypecheck tc_result) tc_warnings mb_old_h
   -- HsSrcFile Module.
   mb_desugar <-
       if ms_mod summary /= gHC_PRIM && hsc_src == HsSrcFile
-      then Just <$> hscDesugar' (ms_location summary) tc_result
+      then Just <$> hscDesugar' (Just summary) (ms_location summary) tc_result
       else pure Nothing
 
   -- Report the warnings from both typechecking and desugar together
@@ -2334,7 +2338,7 @@ hscParsedDecls hsc_env decls = runInteractiveHsc hsc_env $ do
                                       ml_dyn_obj_file = panic "hsDeclsWithLocation:ml_dyn_obj_file",
                                       ml_dyn_hi_file = panic "hsDeclsWithLocation:ml_dyn_hi_file",
                                       ml_hie_file  = panic "hsDeclsWithLocation:ml_hie_file" }
-    ds_result <- hscDesugar' iNTERACTIVELoc tc_gblenv
+    ds_result <- hscDesugar' Nothing iNTERACTIVELoc tc_gblenv
 
     {- Simplify -}
     simpl_mg <- liftIO $ do
diff --git a/compiler/GHC/Driver/Plugins.hs b/compiler/GHC/Driver/Plugins.hs
index 823ecd9..2dbde1f 100644
--- a/compiler/GHC/Driver/Plugins.hs
+++ b/compiler/GHC/Driver/Plugins.hs
@@ -77,6 +77,7 @@ import GHC.Driver.Plugins.External
 import GHC.Unit.Module
 import GHC.Unit.Module.ModIface
 import GHC.Unit.Module.ModSummary
+import GHC.Unit.Module.ModGuts
 
 import GHC.Parser.Errors.Types (PsWarning, PsError)
 
@@ -184,6 +185,10 @@ data Plugin = Plugin {
     -- the loading of the plugin interface. Tools that rely on information from
     -- modules other than the currently compiled one should implement this
     -- function.
+  , desugarResultAction :: [CommandLineOption] -> Maybe ModSummary -> TcGblEnv -> ModGuts -> Hsc ModGuts
+    -- ^ Modify the desugaring (Core) result before it is optimised. This is
+    -- called by "GHC.Driver.Main" after desugaring produces the initial
+    -- 'ModGuts'.
   }
 
 -- Note [Source plugins]
@@ -286,6 +291,7 @@ defaultPlugin = Plugin {
       , typeCheckResultAction = \_ _ -> return
       , spliceRunAction       = \_ -> return
       , interfaceLoadAction   = \_ -> return
+      , desugarResultAction   = \_ _ _ guts -> return guts
     }
 
 
